### DISCUSSION TITLE
Build Extensions and run on windows application

### DISCUSSION THREAD

**Discussion by DrySlime** [simonec73/threatsmanager#71] [Q&A] [answered]
Hello, 

I would like to ask if it is possible to build the extensions with the source code, move the files to the extensions folder of the windows application and  then run it on the windows application. Because i tried that and it didnt work. For example even the amount of files compared between the build version and the .msi file werent the same, here the .msi had more. 

My goal is to write my own extension and test them out. Could you help me find a way to make that happen

Sincerely Chris

**Comment by simonec73** [comment]
There are several reasons why your library is not loaded:

1. Its name may not start with one of the configured prefixes. To address this, please read [the documentation](https://threatsmanager.com/training/intro/start-here/extensions-configuration/).
2. It may not be signed with one of the approved certificates. To address this, please read [the documentation](https://threatsmanager.com/training/intro/start-here/extensions-configuration/).
3. It may not be included in a folder specified among those allowed for the Extensions. Subfolders are also fine. To address this, please read [the documentation](https://threatsmanager.com/training/intro/start-here/extensions-configuration/). Based on what you wrote, I assume this is not the case. Still, I've included it here for completeness.
4. It may not include the following attribute: **[assembly: ExtensionsContainer("2.1.5")]**, where the version specified (2.1.5) must be the same as the engine used by Threats Manager Studio you are using. You can also specify ranges, like **[assembly: ExtensionsContainer("2.1.0", "2.1.10")]**.
5. It may not use one of the required Nuget packages. If you do not have any WinForms dialog, you can use [ThreatsManager.Utilities](https://www.nuget.org/packages/ThreatsManager.Utilities/), otherwise you should use [ThreatsManager.Utilities.WinForms](https://www.nuget.org/packages/ThreatsManager.Utilities.WinForms/). There are other Nuget packages for specialized needs.
6. Your library may not use the version of the .NET Framework required by the last TMS: 4.8.1.
7. You may not have implemented any Extensions. An Extension is a .NET class that implements an interface derived from IExtension and is decorated with the attribute Extension. You can find examples in the GitHub repository, under folder [Samples](https://github.com/simonec73/threatsmanager/tree/master/Samples).

Some partial guidance is collected in the [Wiki](https://github.com/simonec73/threatsmanager/wiki). I admit I must work on it, but still, it's something you can use to start.

I hope it helps.

Simone

**Reply by DrySlime** [reply]
Thank You for this response. I was able to build my really small extension. Now i came across another hurdle, I am not sure if it is a newbie question but regardless, is there a way to debug the code?

 In my scenario i wanted to check the entities but for some reason the program says there are none and here i am now, trying to debug and understand it. I tried to attach the debugger of visual studio but to no avail.
 
 Thank you for your time and effort;
 
 Sincerely
 
 Chris

**Reply by simonec73** [reply, marked as answer]
The latest versions include full debug symbols. Moreover, you can download the sources from GitHub. Normally, you should not be concerned with debugging my code because your focus should be on your code, but having full debug symbols and the sources available should allow you to gain insights into what happens in the engine and in other parts of my code. 

Debugging involves configuring a project to run threatsmanager.exe when you start debugging your library. Then, you include a breakpoint in your code and check what you receive as input to your Extension. If your library is not loaded, you may not have configured TMS to load it. You can just open TMS and configure the location of your debug output folder and the prefix of your library. When debugging, you should turn off certificate checking because you usually sign the libraries only for the code you publish.
